File: BackEnd/src/data/mappers/user_mapper.ts
/**
 * @file user_mapper.ts
 * @description Implementación del mapper de usuarios.
 * Transforma entre UserDocument (capa data) y User (entidad de dominio).
 * Es la única clase que conoce ambas representaciones.
 * @module Auth
 * @layer Data
 *
 * @implements {IUserMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import { IUserMapper } from '../IMappers/IUserMapper.js';
import { User } from '../../domain/entities/User.js';
import { UserDocument } from '../datasources/mongodb/models/UserModel.js';

/**
 * Mapper concreto de usuarios entre capa data y dominio.
 *
 * @implements {IUserMapper}
 * @injectable
 */
@injectable()
export class UserMapper implements IUserMapper {
  /**
   * Convierte un UserDocument de MongoDB en una entidad User de dominio.
   *
   * @param doc — Documento MongoDB con ObjectId y Dates nativos.
   * @returns Entidad User del dominio.
   */
  toEntity(doc: UserDocument): User {
    return new User({
      id:           doc._id.toHexString(),
      role:         doc.role ?? 'user',
      name:         doc.name,
      email:        doc.email,
      passwordHash: doc.passwordHash,
      // Normalizar null a undefined en campos opcionales para mantener la entidad limpia
      photo:        doc.photo       ?? undefined,
      bannerPhoto:  doc.bannerPhoto ?? undefined,
      bio:          doc.bio         ?? undefined,
      location:     doc.location    ?? undefined,
      locationLat:  doc.locationLat ?? undefined,
      locationLon:  doc.locationLon ?? undefined,
      fcmToken:     doc.fcmToken    ?? undefined,
      lastChatPushTitle: doc.lastChatPushTitle ?? null,
      preferences:  doc.preferences,
      createdAt:    doc.createdAt,
      updatedAt:    doc.updatedAt,
      deletedAt:    doc.deletedAt  ?? undefined,
      bannedUntil:  doc.bannedUntil ?? undefined,
    });
  }

  /**
   * Convierte una entidad User de dominio en un documento para MongoDB.
   * No incluye _id (lo asigna el repositorio al insertar).
   *
   * @param user — Entidad de dominio.
   * @returns Documento MongoDB sin _id.
   */
  toDocument(user: User): Omit<UserDocument, '_id'> {
    return {
      role:         user.role,
      name:         user.name,
      email:        user.email,
      passwordHash: user.passwordHash,
      photo:        user.photo,
      bannerPhoto:  user.bannerPhoto,
      bio:          user.bio,
      location:     user.location,
      locationLat:  user.locationLat,
      locationLon:  user.locationLon,
      fcmToken:     user.fcmToken,
      ...(user.lastChatPushTitle !== undefined && { lastChatPushTitle: user.lastChatPushTitle }),
      preferences:  user.preferences,
      createdAt:    user.createdAt,
      updatedAt:    user.updatedAt,
      // Omitir deletedAt cuando es undefined para evitar que el driver almacene null
      ...(user.deletedAt !== undefined && { deletedAt: user.deletedAt }),
      ...(user.bannedUntil !== undefined && { bannedUntil: user.bannedUntil }),
    };
  }
}

File: FrontEnd/lib/data/mappers/user_mapper.dart
/// @file user_mapper.dart
/// @description Implementación del mapper de usuario.
/// Convierte UserModel ↔ User normalizando tipos y valores por defecto.
/// Toda la lógica de transformación vive aquí, nunca en el Model ni en la entidad.
/// @module Core
/// @layer Data
library;

import '../../core/utils/date_utils.dart';
import '../../domain/entities/user.dart';
import '../i_mappers/i_user_mapper.dart';
import '../models/user_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// USER MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Implementación de [IUserMapper].
///
/// [implements] IUserMapper
/// [injectable] registrar en container.dart como singleton.
class UserMapper implements IUserMapper {

  // ─── UserModel → User ─────────────────────────────────────────────────────────

  @override
  User toEntity(UserModel model) {
    return User(
      id:          model.id,
      name:        model.name,
      email:       model.email,
      role:        model.role,
      photo:       model.photo,
      bannerPhoto: model.bannerPhoto,
      bio:         model.bio,
      location:    model.location,
      locationLat: model.locationLat,
      locationLon: model.locationLon,
      fcmToken:    model.fcmToken,
      bannedUntil: model.bannedUntil != null
          ? PlantDateUtils.parseUtc(model.bannedUntil!)
          : null,
      createdAt:   PlantDateUtils.parseUtc(model.createdAt) ?? DateTime.now().toUtc(),
      preferences: model.preferences != null
          ? _toPreferences(model.preferences!)
          : null,
    );
  }

  // ─── User → UserModel ─────────────────────────────────────────────────────────

  @override
  UserModel toModel(User entity) {
    return UserModel(
      id:          entity.id,
      name:        entity.name,
      email:       entity.email,
      role:        entity.role,
      photo:       entity.photo,
      bannerPhoto: entity.bannerPhoto,
      bio:         entity.bio,
      location:    entity.location,
      locationLat: entity.locationLat,
      locationLon: entity.locationLon,
      fcmToken:    entity.fcmToken,
      bannedUntil: entity.bannedUntil != null
          ? PlantDateUtils.toIso8601(entity.bannedUntil!)
          : null,
      createdAt:   PlantDateUtils.toIso8601(entity.createdAt),
      preferences: entity.preferences != null
          ? _toPreferencesModel(entity.preferences!)
          : null,
    );
  }

  // ─── Helpers de preferencias ──────────────────────────────────────────────────

  UserPreferences _toPreferences(UserPreferencesModel model) {
    return UserPreferences(
      pushNotifications:    model.pushNotifications    ?? true,
      profilePublic:        model.profilePublic        ?? true,
      language:             model.language             ?? 'es',
    );
  }

  UserPreferencesModel _toPreferencesModel(UserPreferences prefs) {
    return UserPreferencesModel(
      pushNotifications:    prefs.pushNotifications,
      profilePublic:        prefs.profilePublic,
      language:             prefs.language,
    );
  }
}
